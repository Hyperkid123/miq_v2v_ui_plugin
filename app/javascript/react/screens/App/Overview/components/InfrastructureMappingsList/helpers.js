import Immutable from 'seamless-immutable';
import { networkKey } from '../../../common/networkKey';

export const mapInfrastructureMappings = (transformation_mapping_items, clusters, datastores, networks) => {
  /**
   * map the target source -> destination clusters/networks/datastores for
   * display on the infrastructure mappings list view
   *
   * SQL-JOIN-MIQ-IN-JS ;)
   *
   * results in the following structure:
   * {
   *  targetClusters: {
   *    `clusterId`: {
   *      targetCluster: {...},
   *      sourceClusters: [{...}]
   *    }
   *  },
   *  targetDatastores: {
   *   `datastoreId`: {
   *     target: {
   *       targetCluster: {...},
   *       targetDatastore: {...}
   *     },
   *     sources: [
   *       {
   *         sourceDatastore: {...},
   *         sourceCluster: {...}
   *       }
   *     ]
   *   }
   *  }
   *  targetNetworks: {
   *   `networkId`: {
   *     target: {
   *       targetCluster: {...},
   *       targetNetwork: {...}
   *     },
   *     sources: [
   *       {
   *         sourceNetwork: {...},
   *         sourceCluster: {...}
   *       }
   *     ]
   *   }
   *  }
   * }
   *
   */
  const clusterMappingItems = transformation_mapping_items.filter(item => {
    const destination = item.destination_type.toLowerCase();
    return destination === 'emscluster' || destination === 'cloudtenant';
  });
  const datastoreMappingItems = transformation_mapping_items.filter(item => {
    const destination = item.destination_type.toLowerCase();
    return destination === 'storage' || destination === 'cloudvolumetype';
  });
  const networkMappingItems = transformation_mapping_items.filter(item => {
    const destination = item.destination_type.toLowerCase();
    return destination === 'lan' || destination === 'cloudnetwork';
  });

  // create unique cluster mappings by unique target cluster
  const targetClusters = {};
  for (const clusterMapping of clusterMappingItems) {
    const sourceCluster = clusters.find(c => c.id === clusterMapping.source_id);
    const targetCluster = clusters.find(c => c.id === clusterMapping.destination_id);

    if (sourceCluster && targetCluster) {
      if (targetClusters[targetCluster.id]) {
        targetClusters[targetCluster.id].sourceClusters.push(sourceCluster);
      } else {
        targetClusters[targetCluster.id] = {};
        targetClusters[targetCluster.id].targetCluster = targetCluster;
        targetClusters[targetCluster.id].sourceClusters = [sourceCluster];
      }
    } else {
      // we have missing cluster data, just return null and show missing data
      return {
        targetClusters: null,
        targetDatastores: null,
        targetNetworks: null
      };
    }
  }

  // transform cluster lans and datastores to key/value lookups for use in datastore/lan mappings
  const clusterDatastores = {};
  const clusterLans = {};
  clusters.forEach(cluster => {
    if (cluster.storages && cluster.storages.length) {
      cluster.storages.forEach(datastore => {
        clusterDatastores[datastore.id] = cluster.id;
      });
    }
    if (cluster.lans && cluster.lans.length) {
      cluster.lans.forEach(lan => {
        clusterLans[lan.id] = cluster.id;
      });
    }
  });

  // create unique datastore mappings by unique target datastore
  const targetDatastores = {};
  let missingDatastores = false;
  for (const datastoreMapping of datastoreMappingItems) {
    const sourceCluster = clusters.find(c => c.id === clusterDatastores[datastoreMapping.source_id]);
    const targetCluster = clusters.find(c => c.id === clusterDatastores[datastoreMapping.destination_id]);
    const sourceDatastore = datastores.find(d => d.id === datastoreMapping.source_id);
    const targetDatastore = datastores.find(d => d.id === datastoreMapping.destination_id);

    if (sourceCluster && targetCluster && sourceDatastore && targetDatastore) {
      const source = {
        sourceDatastore,
        sourceCluster
      };
      const target = {
        targetDatastore,
        targetCluster
      };
      if (targetDatastores[targetDatastore.id]) {
        targetDatastores[targetDatastore.id].sources.push(source);
      } else {
        targetDatastores[targetDatastore.id] = {};
        targetDatastores[targetDatastore.id].target = target;
        targetDatastores[targetDatastore.id].sources = [source];
      }
    } else {
      missingDatastores = true;
      break;
    }
  }

  // create unique networks mappings by unique target network
  const targetNetworks = {};
  let missingNetworks = false;
  for (const networkMapping of networkMappingItems) {
    if (!(networkMapping.source_id in clusterLans)) {
      missingNetworks = true;
      break;
    }
    if (!(networkMapping.destination_id in clusterLans)) {
      missingNetworks = true;
      break;
    }
    const sourceCluster = clusters.find(c => c.id === clusterLans[networkMapping.source_id]);
    const targetCluster = clusters.find(c => c.id === clusterLans[networkMapping.destination_id]);
    const sn = networks.find(d => d.id === networkMapping.source_id);
    const tn = networks.find(d => d.id === networkMapping.destination_id);

    if (sourceCluster && targetCluster && sn && tn) {
      const sourceNetwork = Immutable.set(sn, 'clusterId', sourceCluster.id);
      const targetNetwork = Immutable.set(tn, 'clusterId', targetCluster.id);
      const source = {
        sourceNetwork,
        sourceCluster
      };

      const target = {
        targetNetwork,
        targetCluster
      };

      const targetNetworkKey = networkKey(targetNetwork);
      if (targetNetworks[targetNetworkKey]) {
        const duplicatedLanIndex = targetNetworks[targetNetworkKey].sources.findIndex(
          s => networkKey(s.sourceNetwork) === networkKey(sourceNetwork)
        );
        if (duplicatedLanIndex === -1) {
          targetNetworks[targetNetworkKey].sources.push(source);
        }
      } else {
        targetNetworks[targetNetworkKey] = {};
        targetNetworks[targetNetworkKey].target = target;
        targetNetworks[targetNetworkKey].sources = [source];
      }
    } else {
      missingNetworks = true;
      break;
    }
  }

  return {
    targetClusters,
    targetDatastores: missingDatastores ? null : targetDatastores,
    targetNetworks: missingNetworks ? null : targetNetworks
  };
};
