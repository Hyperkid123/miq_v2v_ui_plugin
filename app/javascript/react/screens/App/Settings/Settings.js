import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm, Field } from 'redux-form';
import { Breadcrumb, Form, Grid, Button } from 'patternfly-react';
import Toolbar from '../../../config/Toolbar';
import NumberInput from '../common/forms/NumberInput';

const normalizeStringToInt = str => str && parseInt(str.replace(/\D/g, ''), 10) || 0;

class Settings extends React.Component {
  componentDidMount() {
    const { fetchSettingsAction } = this.props;
    fetchSettingsAction('TODO: url here'); // TODO
  }

  onApplyClick = () => {
    console.log('TODO: apply settings changes');
  };

  render() {
    const { settingsForm } = this.props;

    const toolbarContent = (
      <Toolbar>
        <Breadcrumb.Item href="/dashboard/maintab?tab=compute">{__('Compute')}</Breadcrumb.Item>
        <Breadcrumb.Item href="/migration">{__('Migration')}</Breadcrumb.Item>
        <Breadcrumb.Item active>{__('Migration Settings')}</Breadcrumb.Item>
      </Toolbar>
    );

    const settingsContent = (
      <div className="migration-settings">
        <h2>{__('Concurrent Migrations')}</h2>
        <Form horizontal>
          <Form.FormGroup>
            <Grid.Col componentClass={Form.ControlLabel} sm={5}>
              {__('Maximum Concurrent migrations per conversion host')}
            </Grid.Col>
            <Grid.Col sm={7}>
              <div style={{ width: 100 }}>
                <Field
                  id="maxMigrationsPerHost"
                  name="maxMigrationsPerHost"
                  component={NumberInput}
                  normalize={normalizeStringToInt}
                />
              </div>
            </Grid.Col>
          </Form.FormGroup>
          <Form.FormGroup>
            <Grid.Col sm={12}>
              <Button bsStyle="primary" onClick={this.onApplyClick}>
                {__('Apply')}
              </Button>
            </Grid.Col>
          </Form.FormGroup>
        </Form>
      </div>
    );

    return (
      <React.Fragment>
        {toolbarContent}
        {settingsContent}
      </React.Fragment>
    );
  }
}

Settings.propTypes = {
  fetchSettingsAction: PropTypes.func,
  settingsForm: PropTypes.object
};

export default reduxForm({
  form: 'settings'
})(Settings);
