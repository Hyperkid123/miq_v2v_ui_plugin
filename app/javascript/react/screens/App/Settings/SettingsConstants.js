export const V2V_FETCH_SERVERS = 'V2V_FETCH_SERVERS';
export const V2V_FETCH_SETTINGS = 'V2V_FETCH_SETTINGS';
export const V2V_PATCH_SETTINGS = 'V2V_PATCH_SETTINGS';
export const FETCH_V2V_CONVERSION_HOSTS = 'FETCH_V2V_CONVERSION_HOSTS';
