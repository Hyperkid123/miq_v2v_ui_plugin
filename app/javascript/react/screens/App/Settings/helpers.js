export const getFormValuesFromApiSettings = payload => ({
  max_concurrent_tasks_per_host: payload.transformation.limits.max_concurrent_tasks_per_host
});

export const getApiSettingsFromFormValues = values => ({
  transformation: {
    limits: {
      max_concurrent_tasks_per_host: values.max_concurrent_tasks_per_host
    }
  }
});

// Example task name: "Configuring a conversion_host: operation=enable resource=(type: ManageIQ::Providers::Openstack::CloudManager::Vm id:42000000000113)"
const conversionHostTaskNameRegex = /operation=(\w+)\s+resource=\(type:\s+([\w:]+)\s+id:(\d+)\)/;

export const parseConversionHostTasksMetadata = tasks =>
  tasks &&
  tasks.map(task => {
    const result = conversionHostTaskNameRegex.exec(task.name);
    return {
      ...task,
      nameMeta: {
        operation: result[1],
        resourceType: result[2],
        resourceId: result[3]
      }
    };
  });
