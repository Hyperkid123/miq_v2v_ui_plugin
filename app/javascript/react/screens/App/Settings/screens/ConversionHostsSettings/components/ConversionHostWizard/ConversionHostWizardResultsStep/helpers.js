export const getConfigureConversionHostPostBody = (locationStepValues, hostsStepValues, authStepValues) => {
  console.log('TODO: use location step values:', locationStepValues);
  console.log('TODO: use hosts step values:', hostsStepValues);
  console.log('TODO: use auth step values: ', authStepValues);
  // TODO
  return {
    foo: 'bar'
  };
};
