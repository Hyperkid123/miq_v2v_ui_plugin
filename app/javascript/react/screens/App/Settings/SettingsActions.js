import URI from 'urijs';
// import API from '../../../../common/API';

import { V2V_FETCH_SETTINGS } from './SettingsConstants';

const _getSettingsActionCreator = url => dispatch =>
  dispatch({
    type: V2V_FETCH_SETTINGS,
    // payload: API.get(url)
    // TODO replace this dummy payload with actual settings from API
    payload: new Promise(resolve => {
      setTimeout(
        () =>
          resolve({
            maxMigrationsPerHost: 10
          }),
        2000
      );
    })
  });

export const fetchSettingsAction = url => {
  const uri = new URI(url);
  return _getSettingsActionCreator(uri.toString());
};
