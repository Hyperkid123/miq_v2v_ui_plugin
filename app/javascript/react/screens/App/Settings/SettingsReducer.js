import Immutable from 'seamless-immutable';

import { V2V_FETCH_SETTINGS, V2V_POST_SETTINGS } from './SettingsConstants';

export const initialState = Immutable({
  isFetchingSettings: false,
  fetchingSettingsRejected: false,
  errorFetchingSettings: null,
  savedSettings: {},
  isSavingSettings: false,
  savingSettingsRejected: false,
  errorSavingSettings: null
});

export default (state = initialState, action) => {
  switch (action.type) {
    case `${V2V_FETCH_SETTINGS}_PENDING`:
      return state
        .set('isFetchingSettings', true)
        .set('fetchingSettingsRejected', false)
        .set('errorFetchingSettings', null);
    case `${V2V_FETCH_SETTINGS}_REJECTED`:
      return state
        .set('isFetchingSettings', false)
        .set('fetchingSettingsRejected', true)
        .set('errorFetchingSettings', action.payload);
    case `${V2V_FETCH_SETTINGS}_FULFILLED`:
      return state
        .set('isFetchingSettings', false)
        .set('fetchingSettingsRejected', false)
        .set('errorFetchingSettings', null)
        .set('savedSettings', action.payload);

    case `${V2V_POST_SETTINGS}_PENDING`:
      return state
        .set('isSavingSettings', true)
        .set('savingSettingsRejected', false)
        .set('errorSavingSettings', null);
    case `${V2V_POST_SETTINGS}_REJECTED`:
      return state
        .set('isSavingSettings', false)
        .set('savingSettingsRejected', true)
        .set('errorSavingSettings', action.payload);
    case `${V2V_POST_SETTINGS}_FULFILLED`:
      return state
        .set('isSavingSettings', false)
        .set('savingSettingsRejected', false)
        .set('errorSavingSettings', null)
        .set('savedSettings', action.payload);

    default:
      return state;
  }
};
