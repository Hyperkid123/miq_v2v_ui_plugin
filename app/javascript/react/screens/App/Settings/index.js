import { connect } from 'react-redux';
import Settings from './Settings';

import * as SettingsActions from './SettingsActions';
import * as NotificationActions from '../common/NotificationList/NotificationListActions';

import reducer from './SettingsReducer';

export const reducers = { settings: reducer, form: {} };

const mapStateToProps = ({ settings }, ownProps) => ({
  ...settings,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) => Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  Object.assign(SettingsActions, NotificationActions),
  mergeProps
)(Settings);
