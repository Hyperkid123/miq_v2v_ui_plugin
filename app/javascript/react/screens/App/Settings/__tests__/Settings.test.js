import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Settings from '../Settings';
import { settings } from '../settings.fixures';
import { getFormValuesFromApiSettings } from '../helpers';

describe('Settings component', () => {
  const defaultFormValues = getFormValuesFromApiSettings(settings);
  const baseProps = {
    fetchServersAction: jest.fn(),
    fetchSettingsAction: jest.fn(),
    patchSettingsAction: jest.fn(),
    savedSettings: defaultFormValues,
    settingsForm: { values: defaultFormValues }
  };

  it('renders the settings page', () => {
    const component = shallow(<Settings {...baseProps} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
