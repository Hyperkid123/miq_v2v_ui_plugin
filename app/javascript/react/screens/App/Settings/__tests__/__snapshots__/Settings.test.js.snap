// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component renders the settings page 1`] = `
<Connect(Form(Settings))
  destroyOnUnmount={true}
  enableReinitialize={false}
  fetchServersAction={[MockFunction]}
  fetchSettingsAction={[MockFunction]}
  forceUnregisterOnUnmount={false}
  form="settings"
  getFormState={[Function]}
  keepDirtyOnReinitialize={false}
  patchSettingsAction={[MockFunction]}
  persistentSubmitErrors={false}
  pure={true}
  savedSettings={
    Object {
      "max_concurrent_tasks_per_host": 10,
    }
  }
  settingsForm={
    Object {
      "values": Object {
        "max_concurrent_tasks_per_host": 10,
      },
    }
  }
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;
